Adopt the C++ spaceship operator in more places

cdumez opened the pull request:
#### 49059735c89606ea4750dae483a6cfc28b1b3ef2
<pre>
Adopt the C++ spaceship operator in more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=291247">https://bugs.webkit.org/show_bug.cgi?id=291247</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/CheckedArithmetic.h:
(WTF::Checked::operator&lt;=&gt; const):
(WTF::Checked::operator&lt; const): Deleted.
(WTF::Checked::operator&lt;= const): Deleted.
(WTF::Checked::operator&gt; const): Deleted.
(WTF::Checked::operator&gt;= const): Deleted.
* Source/WTF/wtf/EnumeratedArray.h:
(WTF::EnumeratedArray::operator&lt;=&gt; const):
(WTF::EnumeratedArray::operator&lt; const): Deleted.
(WTF::EnumeratedArray::operator&lt;= const): Deleted.
(WTF::EnumeratedArray::operator&gt; const): Deleted.
(WTF::EnumeratedArray::operator&gt;= const): Deleted.
* Source/WTF/wtf/MediaTime.cpp:
(WTF::operator&lt;=&gt;):
(WTF::MediaTime::compare const): Deleted.
* Source/WTF/wtf/MediaTime.h:
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::operator==):
(WTF::operator&lt;=&gt;):
(WTF::operator&gt;): Deleted.
(WTF::operator&gt;=): Deleted.
(WTF::operator&lt;): Deleted.
(WTF::operator&lt;=): Deleted.
* Source/WTF/wtf/TimeWithDynamicClockType.cpp:
(WTF::operator&lt;=&gt;):
(WTF::TimeWithDynamicClockType::operator&lt; const): Deleted.
(WTF::TimeWithDynamicClockType::operator&gt; const): Deleted.
(WTF::TimeWithDynamicClockType::operator&lt;= const): Deleted.
(WTF::TimeWithDynamicClockType::operator&gt;= const): Deleted.
* Source/WTF/wtf/TimeWithDynamicClockType.h:
* Source/WebCore/Modules/indexeddb/IDBCursor.cpp:
(WebCore::IDBCursor::continueFunction):
* Source/WebCore/Modules/indexeddb/IDBFactory.cpp:
(WebCore::IDBFactory::cmp):
* Source/WebCore/Modules/indexeddb/IDBKey.cpp:
(WebCore::operator&lt;=&gt;):
(WebCore::IDBKey::compare const): Deleted.
(WebCore::IDBKey::isLessThan const): Deleted.
(WebCore::IDBKey::isEqual const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKey.h:
(WebCore::IDBKey::createMultiEntryArray):
(WebCore::IDBKey::operator==):
(WebCore::compareBinaryKeyData):
* Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
(WebCore::operator&lt;=&gt;):
(WebCore::IDBKeyData::compare const): Deleted.
(WebCore::IDBKeyData::operator&lt; const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKeyData.h:
(WebCore::IDBKeyData::operator&gt; const): Deleted.
(WebCore::IDBKeyData::operator&lt;= const): Deleted.
(WebCore::IDBKeyData::operator&gt;= const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKeyRange.cpp:
(WebCore::IDBKeyRange::bound):
(WebCore::IDBKeyRange::isOnlyKey const):
(WebCore::IDBKeyRange::includes):
* Source/WebCore/Modules/indexeddb/IDBKeyRangeData.cpp:
(WebCore::IDBKeyRangeData::isExactlyOneKey const):
(WebCore::IDBKeyRangeData::containsKey const):
* Source/WebCore/Modules/indexeddb/server/IndexValueStore.cpp:
(WebCore::IDBServer::IndexValueStore::lowestIteratorInRange const):
(WebCore::IDBServer::IndexValueStore::highestReverseIteratorInRange const):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
(WebCore::IDBServer::MemoryObjectStore::lowestKeyWithRecordInRange const):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
(WebCore::IDBServer::MemoryObjectStoreCursor::setForwardIteratorFromRemainingRange):
(WebCore::IDBServer::MemoryObjectStoreCursor::setReverseIteratorFromRemainingRange):
(WebCore::IDBServer::MemoryObjectStoreCursor::incrementForwardIterator):
(WebCore::IDBServer::MemoryObjectStoreCursor::incrementReverseIterator):
* Source/WebCore/Modules/indexeddb/server/SQLiteIDBBackingStore.cpp:
(WebCore::IDBServer::idbKeyCollate):
* Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
(WebCore::IDBServer::SQLiteIDBCursor::fetch):
(WebCore::IDBServer::SQLiteIDBCursor::iterate):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::injectIDBKeyIntoScriptValue):

Canonical link: <a href="https://commits.webkit.org/293455@main">https://commits.webkit.org/293455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eeafde9260eefc1136fc10995b285dc99b428bf8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98978 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8859 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104106 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49570 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18910 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27065 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75349 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32477 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14372 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89387 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55710 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14164 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7362 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48947 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91668 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84094 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106473 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97608 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26075 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19007 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84315 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26450 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85584 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83818 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21249 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28473 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6139 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Skipped layout-tests; 10 flakes") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19802 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26028 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31214 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121224 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25848 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33893 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27422 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.cpp:
```diff
@@ -107,54 +107,59 @@ bool IDBKey::isValid() const
-        return -codePointCompare(std::get<String>(other.m_value), std::get<String>(m_value));
+    case IndexedDB::KeyType::String: {
+        // FIXME: codePointCompare() should return a std::strong_ordering.
+        int comparison = -codePointCompare(std::get<String>(other.m_value), std::get<String>(m_value));
```
```suggestion
        int comparison = codePointCompare(std::get<String>(m_value), std::get<String>(other.m_value));
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.cpp:
```diff
@@ -107,54 +107,59 @@ bool IDBKey::isValid() const
         auto number = std::get<double>(m_value);
         auto otherNumber = std::get<double>(other.m_value);
-        return (number < otherNumber) ? -1 : ((number > otherNumber) ? 1 : 0);
+        return (number < otherNumber) ? std::weak_ordering::less : ((number > otherNumber) ? std::weak_ordering::greater : std::weak_ordering::equivalent);
```
```suggestion
        return number <=> otherNumber;
```

cdumez replied:
I think this doesn't work because <=> returns a partial_ordering for double, and this function returns a weak_ordering.

darinadler replied:
Some thoughts on how to do this elegantly: Seems like we could have a helper function that converts a `std::partial_ordering` to a `std::weak_ordering`, maybe with a `RELEASE_ASSERT` that it’s not unordered. Or we could make a function that does the equivalent of `<=>` on two doubles but returns a `std::weak_ordering`, with a similar assertion.

cdumez replied:
Ok, maybe this then:
```
template<typename FloatingPointType>
ALWAYS_INLINE std::weak_ordering compareFloatingPointWithWeakOrdering(FloatingPointType a, FloatingPointType b) requires (std::is_floating_point_v<FloatingPointType>)
{
    return (a < b) ? std::weak_ordering::less : ((a > b) ? std::weak_ordering::greater : std::weak_ordering::equivalent);
}
```

darinadler replied:
That function has a long name, which is annoying since it's only a slight variation on `<=>`. Also, if either value is `NaN` it returns `equivalent` silently without complaint, which does not seem ideal. Maybe we should have a function that is a checked cast to weak ordering. Almost think we should make `downcast<std::weak_ordering>(a <=> b)` work. Or we could have `weakOrderingCast` or `weak_ordering_cast`. Seems valuable to either assert neither value is `NaN` or assert it's not unordered.

cdumez replied:
Ok, maybe this then:
```
ALWAYS_INLINE std::weak_ordering weakOrderingCast(std::partial_ordering ordering)
{
    RELEASE_ASSERT(ordering != std::partial_ordering::unordered);
    if (is_eq(ordering))
        return std::weak_ordering::equivalent;
    return is_lt(ordering) ? std::weak_ordering::less : std::weak_ordering::greater;
}

return weakOrderingCast(std::get<double>(m_value) <=> std::get<double>(other.m_value));
```

I don't want to use downcast because this is not truly a down cast.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/43846.

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.cpp:
```diff
@@ -107,54 +107,59 @@ bool IDBKey::isValid() const
-            return 1;
-        return 0;
+            return std::weak_ordering::greater;
+        return std::weak_ordering::equivalent;
```
```suggestion
        return array.size() <=> otherArray.size();
```

cdumez replied:
Nice!

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.h:
```diff
@@ -172,39 +172,39 @@ class IDBKey : public RefCounted<IDBKey> {
+            return std::strong_ordering::greater;
         if (a[i] < b[i])
-            return -1;
+            return std::strong_ordering::less;
```
```suggestion
        if (auto result = (a[i] <=> b[i]); is_ne(result))
            return result;
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.h:
```diff
@@ -172,39 +172,39 @@ class IDBKey : public RefCounted<IDBKey> {
+        return std::strong_ordering::greater;
 
-    return -1;
+    return std::strong_ordering::less;
```
```suggestion
    return a.size() <=> b.size();
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.h:
```diff
@@ -172,39 +172,39 @@ class IDBKey : public RefCounted<IDBKey> {
-        return 0;
+        return std::strong_ordering::equal;
 
     if (aData && !bData)
```
```suggestion
    if (!bData)
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.h:
```diff
@@ -172,39 +172,39 @@ class IDBKey : public RefCounted<IDBKey> {
     if (aData && !bData)
-        return 1;
+        return std::strong_ordering::greater;
     if (!aData && bData)
```
```suggestion
    if (!aData)
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
```diff
@@ -249,69 +249,74 @@ bool IDBKeyData::decode(KeyedDecoder& decoder, IDBKeyData& result)
-    if (type != otherType)
-        return type < otherType ? 1 : -1;
+    if (aType != bType)
+        return aType < bType ? std::weak_ordering::greater : std::weak_ordering::less;
```
```suggestion
        return bType <=> aType;
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
```diff
@@ -249,69 +249,74 @@ bool IDBKeyData::decode(KeyedDecoder& decoder, IDBKeyData& result)
+            return std::weak_ordering::less;
+        if (aArray.size() > bArray.size())
+            return std::weak_ordering::greater;
+        return std::weak_ordering::equivalent;
```
```suggestion
        return aArray.size() <=> bArray.size();
```

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
```diff
@@ -249,69 +249,74 @@ bool IDBKeyData::decode(KeyedDecoder& decoder, IDBKeyData& result)
-        return number > otherNumber ? 1 : -1;
+        if (aNumber == bNumber)
+            return std::weak_ordering::equivalent;
+        return aNumber > bNumber ? std::weak_ordering::greater : std::weak_ordering::less;
```
```suggestion
        return aNumber <=> bNumber;
```

cdumez replied:
I do not believe this works because those are `double` and <=> returns a partial_ordering for those. Please let me know if you still think there is a way.

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
```diff
@@ -249,69 +249,74 @@ bool IDBKeyData::decode(KeyedDecoder& decoder, IDBKeyData& result)
-        return number > otherNumber ? 1 : -1;
+        if (aNumber == bNumber)
+            return std::weak_ordering::equivalent;
+        return aNumber > bNumber ? std::weak_ordering::greater : std::weak_ordering::less;
```
```suggestion
        return aNumber <=> bNumber;
```

cdumez replied:
ditto.

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKeyRangeData.cpp:
```diff
@@ -60,23 +60,23 @@ bool IDBKeyRangeData::isExactlyOneKey() const
         return false;
 
-    return !lowerKey.compare(upperKey);
+    return is_eq(lowerKey <=> upperKey);
```
How is this different from using `==`?

darinadler commented on Source/WebCore/Modules/indexeddb/server/IndexValueStore.cpp:
```diff
@@ -198,7 +198,7 @@ IDBKeyDataSet::iterator IndexValueStore::lowestIteratorInRange(const IDBKeyRange
 
     if (!range.upperKey.isNull()) {
-        if (lowestInRange->compare(range.upperKey) > 0)
+        if (is_gt(*lowestInRange <=> range.upperKey))
```
Can we use `>` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/IndexValueStore.cpp:
```diff
@@ -222,7 +222,7 @@ IDBKeyDataSet::reverse_iterator IndexValueStore::highestReverseIteratorInRange(c
 
     if (!range.lowerKey.isNull()) {
-        if (highestInRange->compare(range.lowerKey) < 0)
+        if (is_lt(*highestInRange <=> range.lowerKey))
```
Can we use `<` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
```diff
@@ -504,7 +504,7 @@ IDBKeyData MemoryObjectStore::lowestKeyWithRecordInRange(const IDBKeyRangeData&
 
     if (!keyRangeData.upperKey.isNull()) {
-        if (lowestInRange->compare(keyRangeData.upperKey) > 0)
+        if (is_gt(*lowestInRange <=> keyRangeData.upperKey))
```
Can we use `>` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
```diff
@@ -126,7 +126,7 @@ void MemoryObjectStoreCursor::setForwardIteratorFromRemainingRange(IDBKeyDataSet
 
     if (!m_remainingRange.upperKey.isNull()) {
-        if (lowest->compare(m_remainingRange.upperKey) > 0)
+        if (is_gt(*lowest <=> m_remainingRange.upperKey))
```
Can we use `>` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
```diff
@@ -177,7 +177,7 @@ void MemoryObjectStoreCursor::setReverseIteratorFromRemainingRange(IDBKeyDataSet
 
     if (!m_remainingRange.lowerKey.isNull()) {
-        if (highest->compare(m_remainingRange.lowerKey) < 0)
+        if (is_lt(*highest <=> m_remainingRange.lowerKey))
```
Can we use `<` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
```diff
@@ -233,7 +233,7 @@ void MemoryObjectStoreCursor::incrementForwardIterator(IDBKeyDataSet& set, const
             return;
 
-        if ((*m_iterator)->compare(key) < 0) {
+        if (is_lt(**m_iterator <=> key)) {
```
Can we use `<` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
```diff
@@ -247,7 +247,7 @@ void MemoryObjectStoreCursor::incrementForwardIterator(IDBKeyDataSet& set, const
     // If the forwardIterator was reset because it's previous record had been deleted,
     // we might have already advanced past the current position, eating one one of the iteration counts.
-    if (didResetIterator && (*m_iterator)->compare(m_currentPositionKey) > 0)
+    if (didResetIterator && is_gt(**m_iterator <=> m_currentPositionKey))
```
Can we use `>` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
```diff
@@ -287,7 +287,7 @@ void MemoryObjectStoreCursor::incrementReverseIterator(IDBKeyDataSet& set, const
             return;
 
-        if ((*m_iterator)->compare(key) > 0) {
+        if (is_gt(**m_iterator <=> key)) {
```
Can we use `>` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
```diff
@@ -302,7 +302,7 @@ void MemoryObjectStoreCursor::incrementReverseIterator(IDBKeyDataSet& set, const
     // If the reverseIterator was reset because it's previous record had been deleted,
     // we might have already advanced past the current position, eating one one of the iteration counts.
-    if (didResetIterator && (*m_iterator)->compare(m_currentPositionKey) < 0)
+    if (didResetIterator && is_lt(**m_iterator <=> m_currentPositionKey))
```
Can we use `<` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
```diff
@@ -442,7 +442,7 @@ bool SQLiteIDBCursor::fetch()
         m_fetchedRecordsSize += m_fetchedRecords.last().record.size();
 
-        if (m_currentKeyForUniqueness.compare(m_fetchedRecords.last().record.key))
+        if (is_neq(m_currentKeyForUniqueness <=> m_fetchedRecords.last().record.key))
```
Can we use `!=` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
```diff
@@ -598,24 +598,24 @@ bool SQLiteIDBCursor::iterate(const IDBKeyData& targetKey, const IDBKeyData& tar
         // Search for the next key >= the target if the cursor is a Next cursor, or the next key <= if the cursor is a Previous cursor.
         if (m_cursorDirection == IndexedDB::CursorDirection::Next || m_cursorDirection == IndexedDB::CursorDirection::Nextunique) {
-            if (m_fetchedRecords.first().record.key.compare(targetKey) >= 0)
+            if (is_gteq(m_fetchedRecords.first().record.key <=> targetKey))
```
Can we use `>=` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
```diff
@@ -598,24 +598,24 @@ bool SQLiteIDBCursor::iterate(const IDBKeyData& targetKey, const IDBKeyData& tar
+            if (is_gteq(m_fetchedRecords.first().record.key <=> targetKey))
                 break;
-        } else if (m_fetchedRecords.first().record.key.compare(targetKey) <= 0)
+        } else if (is_lteq(m_fetchedRecords.first().record.key <=> targetKey))
```
Can we use `<=` here?

weinig commented on Source/WTF/wtf/MediaTime.h:
```diff
@@ -75,21 +75,11 @@ class WTF_EXPORT_PRIVATE MediaTime final {
-
-    ComparisonFlags compare(const MediaTime& rhs) const;
+    WTF_EXPORT_PRIVATE friend std::weak_ordering operator<=>(const MediaTime&, const MediaTime&);
+    friend bool operator==(const MediaTime& a, const MediaTime& b) { return is_eq(a <=> b); }
```
What's the benefit of using the friend version of these rather than the member version?

cdumez replied:
If I remember correctly the benefits is when you do `a == b` it works even if a is not a MediaTime (but a type convertible to a MediaTime). I remember Darin recommended we prefer this friend version when we started defaulting operator==() in a lot of places so I'm following the same pattern for the spaceship operator.

darinadler commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
```diff
@@ -598,24 +598,24 @@ bool SQLiteIDBCursor::iterate(const IDBKeyData& targetKey, const IDBKeyData& tar
 
     if (targetPrimaryKey.isValid()) {
-        while (!m_fetchedRecords.first().isTerminalRecord() && !m_fetchedRecords.first().record.key.compare(targetKey)) {
+        while (!m_fetchedRecords.first().isTerminalRecord() && is_eq(m_fetchedRecords.first().record.key <=> targetKey)) {
```
Can we use `==` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
```diff
@@ -598,24 +598,24 @@ bool SQLiteIDBCursor::iterate(const IDBKeyData& targetKey, const IDBKeyData& tar
             // Search for the next primary key >= the primary target if the cursor is a Next cursor, or the next key <= if the cursor is a Previous cursor.
             if (m_cursorDirection == IndexedDB::CursorDirection::Next || m_cursorDirection == IndexedDB::CursorDirection::Nextunique) {
-                if (m_fetchedRecords.first().record.primaryKey.compare(targetPrimaryKey) >= 0)
+                if (is_gteq(m_fetchedRecords.first().record.primaryKey <=> targetPrimaryKey))
```
Can we use `>=` here?

darinadler commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
```diff
@@ -598,24 +598,24 @@ bool SQLiteIDBCursor::iterate(const IDBKeyData& targetKey, const IDBKeyData& tar
+                if (is_gteq(m_fetchedRecords.first().record.primaryKey <=> targetPrimaryKey))
                     break;
-            } else if (m_fetchedRecords.first().record.primaryKey.compare(targetPrimaryKey) <= 0)
+            } else if (is_lteq(m_fetchedRecords.first().record.primaryKey <=> targetPrimaryKey))
```
Can we use `<=` here?

darinadler commented on Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
```diff
@@ -347,7 +347,7 @@ bool injectIDBKeyIntoScriptValue(JSGlobalObject& lexicalGlobalObject, const IDBK
     // Do not set if object already has the correct property value.
     JSValue existingKey;
-    if (get(lexicalGlobalObject, parent, keyPathElements.last(), existingKey) && !key->compare(createIDBKeyFromValue(lexicalGlobalObject, existingKey)))
+    if (get(lexicalGlobalObject, parent, keyPathElements.last(), existingKey) && is_eq(key->compare(createIDBKeyFromValue(lexicalGlobalObject, existingKey))))
```
Can we use `==` here?

cdumez committed:
Adopt the C++ spaceship operator in more places
https://bugs.webkit.org/show_bug.cgi?id=291247

Reviewed by Darin Adler.

* Source/WTF/wtf/CheckedArithmetic.h:
(WTF::Checked::operator<=> const):
(WTF::Checked::operator< const): Deleted.
(WTF::Checked::operator<= const): Deleted.
(WTF::Checked::operator> const): Deleted.
(WTF::Checked::operator>= const): Deleted.
* Source/WTF/wtf/EnumeratedArray.h:
(WTF::EnumeratedArray::operator<=> const):
(WTF::EnumeratedArray::operator< const): Deleted.
(WTF::EnumeratedArray::operator<= const): Deleted.
(WTF::EnumeratedArray::operator> const): Deleted.
(WTF::EnumeratedArray::operator>= const): Deleted.
* Source/WTF/wtf/MediaTime.cpp:
(WTF::operator<=>):
(WTF::MediaTime::compare const): Deleted.
* Source/WTF/wtf/MediaTime.h:
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::operator==):
(WTF::operator<=>):
(WTF::operator>): Deleted.
(WTF::operator>=): Deleted.
(WTF::operator<): Deleted.
(WTF::operator<=): Deleted.
* Source/WTF/wtf/TimeWithDynamicClockType.cpp:
(WTF::operator<=>):
(WTF::TimeWithDynamicClockType::operator< const): Deleted.
(WTF::TimeWithDynamicClockType::operator> const): Deleted.
(WTF::TimeWithDynamicClockType::operator<= const): Deleted.
(WTF::TimeWithDynamicClockType::operator>= const): Deleted.
* Source/WTF/wtf/TimeWithDynamicClockType.h:
* Source/WebCore/Modules/indexeddb/IDBCursor.cpp:
(WebCore::IDBCursor::continueFunction):
* Source/WebCore/Modules/indexeddb/IDBFactory.cpp:
(WebCore::IDBFactory::cmp):
* Source/WebCore/Modules/indexeddb/IDBKey.cpp:
(WebCore::operator<=>):
(WebCore::IDBKey::compare const): Deleted.
(WebCore::IDBKey::isLessThan const): Deleted.
(WebCore::IDBKey::isEqual const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKey.h:
(WebCore::IDBKey::createMultiEntryArray):
(WebCore::IDBKey::operator==):
(WebCore::compareBinaryKeyData):
* Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
(WebCore::operator<=>):
(WebCore::IDBKeyData::compare const): Deleted.
(WebCore::IDBKeyData::operator< const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKeyData.h:
(WebCore::IDBKeyData::operator> const): Deleted.
(WebCore::IDBKeyData::operator<= const): Deleted.
(WebCore::IDBKeyData::operator>= const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBKeyRange.cpp:
(WebCore::IDBKeyRange::bound):
(WebCore::IDBKeyRange::isOnlyKey const):
(WebCore::IDBKeyRange::includes):
* Source/WebCore/Modules/indexeddb/IDBKeyRangeData.cpp:
(WebCore::IDBKeyRangeData::isExactlyOneKey const):
(WebCore::IDBKeyRangeData::containsKey const):
* Source/WebCore/Modules/indexeddb/server/IndexValueStore.cpp:
(WebCore::IDBServer::IndexValueStore::lowestIteratorInRange const):
(WebCore::IDBServer::IndexValueStore::highestReverseIteratorInRange const):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
(WebCore::IDBServer::MemoryObjectStore::lowestKeyWithRecordInRange const):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStoreCursor.cpp:
(WebCore::IDBServer::MemoryObjectStoreCursor::setForwardIteratorFromRemainingRange):
(WebCore::IDBServer::MemoryObjectStoreCursor::setReverseIteratorFromRemainingRange):
(WebCore::IDBServer::MemoryObjectStoreCursor::incrementForwardIterator):
(WebCore::IDBServer::MemoryObjectStoreCursor::incrementReverseIterator):
* Source/WebCore/Modules/indexeddb/server/SQLiteIDBBackingStore.cpp:
(WebCore::IDBServer::idbKeyCollate):
* Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
(WebCore::IDBServer::SQLiteIDBCursor::fetch):
(WebCore::IDBServer::SQLiteIDBCursor::iterate):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::injectIDBKeyIntoScriptValue):

Canonical link: https://commits.webkit.org/293455@main
